pyrdp/player/gdi/raster.py

```diff
@@ -0,0 +1,390 @@
+    NOTSRCCOPY: QPainter.RasterOp_NotSource,
+    SRCERASE: QPainter.RasterOp_SourceAndNotDestination,
+    PDna: QPainter.RasterOp_SourceAndNotDestination,
+    DSTINVERT: QPainter.RasterOp_NotDestination,
+    PATINVERT: QPainter.RasterOp_SourceXorDestination,
+    DPan: QPainter.RasterOp_NotSourceOrNotDestination,
+    SRCINVERT: QPainter.RasterOp_SourceXorDestination,
+    DSan: QPainter.RasterOp_NotSourceOrNotDestination,
+    SRCAND: QPainter.RasterOp_SourceAndDestination,
+    DSxn: QPainter.RasterOp_NotSourceXorDestination,
+    DPa: QPainter.RasterOp_SourceAndDestination,
+    PDxn: QPainter.RasterOp_NotSourceXorDestination,
+    DSTCOPY: QPainter.CompositionMode_Destination,
+    DPno: QPainter.RasterOp_NotSourceOrDestination,
+    MERGEPAINT: QPainter.RasterOp_NotSourceOrDestination,
+    SRCCOPY: QPainter.CompositionMode_Source,
+    SDno: QPainter.RasterOp_SourceOrNotDestination,
+    SRCPAINT: QPainter.RasterOp_SourceOrDestination,
+    PATCOPY: QPainter.CompositionMode_Source,
+    PDno: QPainter.RasterOp_SourceOrNotDestination,
+    DPo: QPainter.RasterOp_SourceOrDestination,
+    WHITENESS: QPainter.RasterOp_SetDestination,
+    PSDPxax: QPainter.RasterOp_SourceAndDestination,
+}
+
+
+def set_rop3(op: int, painter: QPainter):
+    """
+    Configure a QPainter with the required ternary raster operation.
+
+    :parma op: The operation identifier.
+    :param painter: The painter being used.
+    :returns: The operation that will be processed.
+    """
+    if op not in _rop3:
+        return None
+
+    mode = _rop3[op]
+    painter.setCompositionMode(mode)
+    return mode
+
+
+def set_rop2(op: int, painter: QPainter):
+    """
+    Configure a QPainter with the required binary raster operation.
+
+    :parma op: The operation identifier.
+    :param painter: The painter being used.
+    :returns: The operation that will be processed.
+    """
+    if op < 0 or op >= len(_rop2):
+        return None
+    mode = _rop2[op]
+    painter.setCompositionMode(mode)
+    return mode
+
+
+def rop_slow(code: str, dst, src, pal):
+    """
+    Slow but generic fallback implementation of raster operations.
+
+    This function implements the RPN-notation described in [MS-RDPEGDI][1]
+    with a generic stack machine. It is much slower than having a hardcoded
+    and optimized function for a particular operation, but greatly reduces
+    the amount of code required.
+
+    [1]: https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpegdi/a9a85075-e796-45eb-b84a-f399324a1109
+    """
+
+    stack = []
+    for c in code:
+        if c == 'D':
+            stack.append(dst)
+        elif c == 'S':
+            stack.append(src)
+        elif c == 'P':
+            if pal is None:
+                raise SyntaxError('Palette is not present.')
+            stack.append(pal)
+        else:
+            lhs = stack.pop()
+            rhs = None
+            res = lhs  # TODO: Actually perform the operation.
```

Comment:
what does this TODO task mean? This isn't implemented?

Reply:
Yes. It ended up not being necessary.